src/box/lua/config/source/file.lua

```diff
@@ -47,6 +44,57 @@ local function new()
     }, mt)
 end
 
+function methods.sync(self, config_module, _iconfig)
+    assert(config_module._config_file ~= nil)
+
+    local sources = {config_module._config_file}
+    local processed_sources = {}
+    local cconfig = {}
+
+    local config
+    local i = 1
+    while i <= #sources do
+        local source = sources[i]
+
+        -- Prevent looping.
+        if processed_sources[source] then
+            log.warn('skipping already processed config file: %q', source)
+            goto continue
+        end
+        processed_sources[source] = true
+
+        config = read_config_file(source)
+
+        config = cluster_config:apply_conditional(config)
+
+        if config.include ~= nil then
+            local insert_idx = 1
+            for _, glob in ipairs(config.include) do
+                local paths = fio.glob(glob)
+                for _, path in ipairs(paths) do
+                    path = fio.abspath(path)
+                    table.insert(sources, i + insert_idx, path)
+                    insert_idx = insert_idx + 1
```

Comment:
Nit: Ugh, this is very mathematic. Here we implement deep-first-search, but the pattern with a queue better serves breadth-first-search. And we have triple loop nesting here.

It would look a bit simpler if we delete processed paths from the queue. So, we just add the new paths to the beginning (or, well, put and pick them from the end, stack alike). It also simplifies the stop condition to just 'if it is empty'.

I think that DFS looks better with recursive implementation (that pass needed context down). However, I don't insist neither on the recursive implementation, nor on getting rid of the index arithmetic. Up to you.

Reply:
Reworked as a DFS on stack. There is a slight complexity, because we need to preserve the order of elements in the `include` array, and can't just iterate over it, pushing the values on stack (it would reverse the order the values are processed in). I tried to comment everything, hope it helps.

Reply:
I like how it looks now, thanks!